modules/ardrivo/test/Arduino.cxx

```diff
@@ -0,0 +1,268 @@
+
+#include <catch2/catch.hpp>
+#include <iostream>
+#include <vector>
+#include <range/v3/algorithm/equal.hpp>
+#include <range/v3/algorithm/for_each.hpp>
+#include "Entrypoint.hxx"
+#include "BoardDataDef.hxx"
+#include "Arduino.h"
+
+using namespace std::literals;
+
+void static init_fake() 
+{
+    if (board_data)
+        (delete board_data), board_data = nullptr;
+    if (board_info)
+        (delete board_info), board_data = nullptr;
+
+    auto loc_board_data = std::make_unique<BoardData>();
+    loc_board_data->silence_errors = false;
+    loc_board_data->start_time = std::chrono::steady_clock::now();
+    loc_board_data->pin_modes = std::vector<std::atomic_uint8_t>(3);
+    loc_board_data->digital_pin_values = std::vector<std::atomic_bool>(3);
+    loc_board_data->analog_pin_values = std::vector<std::atomic_uint16_t>(3);
+    loc_board_data->pin_frequency = std::vector<std::atomic_uint8_t>(3);
+    loc_board_data->interrupts_handlers = std::vector<std::pair<void (*)(), int>>(3);
+    auto loc_board_info = std::make_unique<BoardInfo>();
+    loc_board_info->pins_caps.resize(3);
+    ranges::for_each(loc_board_info->pins_caps, [](PinCapability& pin) {
+        pin.digital_in = pin.digital_out = true;
+        pin.analog_in = pin.analog_out = true;
+        pin.interrupt_able = pin.interrupt_able = true;
+    });
+
+    init(loc_board_data.release(), loc_board_info.release());
+}
+
+TEST_CASE("Read and write digital & analog & pinMode", "[pinMode], [digitalWrite], [digitalRead], [analogWrite], [analogRead]") {
+    init_fake();
+
+    SECTION("Set pinModes", "[pinMode]") 
+    {         
+        auto i = GENERATE(0, 1, 2);
+        board_data->pin_modes[i] = static_cast<uint8_t>(PinMode::OUTPUT);
+        REQUIRE(board_data->pin_modes[i] == 1U);    
+    }
+
+    SECTION("Set digital pins ") 
+    {
+        digitalWrite(1, LOW);
+        REQUIRE(digitalRead(1) == LOW);
+        digitalWrite(1, HIGH);
+        REQUIRE(digitalRead(1) == HIGH);
+
+        auto i = GENERATE(0, 1, 2);
+        digitalWrite(i, HIGH);
+        REQUIRE(digitalRead(i) == HIGH);     
+    }
+    SECTION("Set analog pins") {
+        analogWrite(1, 0);
+        REQUIRE(analogRead(1) == 0);
+        analogWrite(1, 500);
+        REQUIRE(analogRead(1) == 500);
+    }
```

Comment:
I am curious as to how you manage to have a pin in both input and output mode at the same time